tests/steps/Microsoft.DotNet.UpgradeAssistant.Steps.Backup.Tests/BackupStepTests.cs

```diff
@@ -0,0 +1,134 @@
+        [InlineData("B\\B.sln", "B\\Project1\\B.csproj", false, "B.backup", "B.backup\\Project1")]
+        [InlineData("B\\B.csproj", "B\\B.csproj", true, "B.backup", "B.backup")]
+        [InlineData("B\\B.sln", "B\\B.csproj", true, "B.backup", "B.backup\\B.0")]
+        public async Task ApplyTests(string inputPath, string projectPath, bool backupComplete, string expectedBaseBackupPath, string expectedBackupPath)
+        {
+            // Arrange
+            inputPath = Path.GetFullPath(Path.Combine("TestAssets", inputPath));
+            projectPath = Path.GetFullPath(Path.Combine("TestAssets", projectPath));
+            expectedBaseBackupPath = Path.GetFullPath(Path.Combine("TestAssets", expectedBaseBackupPath));
+            expectedBackupPath = Path.GetFullPath(Path.Combine("TestAssets", expectedBackupPath));
+
+            using var mock = GetMock(GetDefaultNonInteractiveOptions(), inputPath, projectPath);
+            var context = mock.Mock<IUpgradeContext>();
+            var step = mock.Create<BackupStep>();
+            step.SetStatus(UpgradeStepStatus.Incomplete);
+            var expectedStatus = UpgradeStepStatus.Complete;
+            var expectedStatusDetails = backupComplete
+                ? $"Backup already exists at {expectedBackupPath}; nothing to do"
+                : $"Project backed up to {expectedBackupPath}";
+
+            // Act
+            try
+            {
+                await step.ApplyAsync(context.Object, CancellationToken.None).ConfigureAwait(true);
+                var status = step.Status;
+                var statusDetails = step.StatusDetails;
+
+                // Assert
+                Assert.Equal(expectedStatus, step.Status);
+                Assert.Equal(expectedStatusDetails, step.StatusDetails);
+
+                var properties = mock.Container.Resolve<IUpgradeContextProperties>();
+                Assert.Single(properties.GetAllPropertyValues());
+                Assert.Equal(expectedBaseBackupPath, properties.GetPropertyValue("BaseBackupLocation"));
+            }
+            finally
+            {
+                if (!backupComplete)
+                {
+                    Directory.Delete(expectedBackupPath, true);
+                }
+            }
+        }
+
+        private static UpgradeOptions GetDefaultNonInteractiveOptions() =>
+            new UpgradeOptions
+            {
+                NonInteractive = true,
+                NonInteractiveWait = 0,
+                SkipBackup = false,
+            };
+
+        private static AutoMock GetMock(UpgradeOptions options, string inputPath, string? projectPath)
+        {
+            var mock = AutoMock.GetLoose(cfg =>
+            {
+                cfg.RegisterInstance(options);
+                cfg.RegisterInstance<IUpgradeContextProperties>(new UpgradeContextProperties());
+                cfg.RegisterType<NonInteractiveUserInput>().As<IUserInput>();
```

Comment:
Do you need to use the `NonInteractiveUserInput` or just mock `IUserInput` to return that it's non-interactive?

Reply:
Good point.

Reply:
Actually, I just checked. BackupStep uses ChooseAsync and AskUserAsync (rather than IsInteractive), so it will be simpler just to use our existing NonInteractiveUserInput type here.